core/include/traccc/utils/type_traits.hpp

```diff
@@ -0,0 +1,32 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2021-2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+namespace traccc::details {
+
+/// Helper trait for detecting when a type is a non-const version of another
+///
+/// This comes into play multiple times to enable certain constructors
+/// conditionally through SFINAE.
+///
+template <typename CTYPE, typename NCTYPE>
+struct is_same_nc {
+    static constexpr bool value = false;
+};
+
+template <typename TYPE>
+struct is_same_nc<TYPE, TYPE> {
+    static constexpr bool value = true;
+};
+
+template <typename TYPE>
+struct is_same_nc<const TYPE, TYPE> {
+    static constexpr bool value = true;
+};
```

Comment:
I would be in favour of getting rid of this entire file and just writing `std::is_same_v<std::decay_t<T1>, std::std::decay_t<T2>>` instead.

Reply:
That would not give the same result... There is a good reason why we have this same setup in vecmem as well. As you may remember, we couldn't find a more standard way for expressing that a type should either be the same as another one, or the constant version of it.

I.e. your proposal would also allow a non-const view to be constructed from a const object.